attest/eventlog.go

```diff
@@ -0,0 +1,399 @@
+package attest
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha1"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"sort"
+
+	// Ensure hashes are available.
+	_ "crypto/sha256"
+
+	"github.com/google/go-tpm/tpm2"
+	"github.com/google/go-tpm/tpmutil"
+)
+
+// TPM algorithms. See the TPM 2.0 specification section 6.3.
+//
+// https://trustedcomputinggroup.org/wp-content/uploads/TPM-Rev-2.0-Part-2-Structures-01.38.pdf#page=42
+const (
+	algSHA1   uint16 = 0x0004
+	algSHA256 uint16 = 0x000B
+)
+
+// EventType indicates what kind of data an event is reporting.
+type EventType uint32
+
+// Event is a single event from a TCG event log. This reports descrete items such
+// as BIOs measurements or EFI states.
+type Event struct {
+	// PCR index of the event.
+	Index int
+	// Type of the event.
+	Type EventType
+
+	// Data of the event. For certain kinds of events, this must match the event
+	// digest to be valid.
+	Data []byte
+	// Digest is the verified digest of the event data. While an event can have
+	// multiple for different hash values, this is the one that was matched to the
+	// PCR value.
+	Digest []byte
+
+	// TODO(ericchiang): Provide examples or links for which event types must
+	// match their data to their digest.
+}
+
+// EventLog contains the data required to parse and validate an event log.
+type EventLog struct {
+	// AIKPublic is the activated public key that has been proven to be under the
+	// control of the TPM.
+	AIKPublic crypto.PublicKey
+	// AIKHash is the hash used to generate the quote.
+	AIKHash crypto.Hash
+
+	// Quote is a signature over the values of a PCR.
+	Quote *Quote
+	// PCRs are the hash values in a given number of registers.
+	PCRs []PCR
+	// Nonce is additional data used to validate the quote signature. It's used
+	// by the server to prevent clients from re-playing quotes.
+	Nonce []byte
+
+	// MeasurementLog contains the raw event log data, which is matched against
+	// the PCRs for validation.
+	MeasurementLog []byte
+}
+
+// Validate verifies the signature of the quote agains the public key, that the
+// quote matches the PCRs, parses the measurement log, and replays the PCRs.
+//
+// Events for PCRs not in the quote are dropped.
+func (e *EventLog) Validate() (events []Event, err error) {
+	var pcrs []PCR
+	switch e.Quote.Version {
+	case TPMVersion12:
+		pcrs, err = e.validate12Quote()
+	case TPMVersion20:
+		pcrs, err = e.validate20Quote()
+	default:
+		return nil, fmt.Errorf("quote used unknown tpm version 0x%x", e.Quote.Version)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("invalid quote: %v", err)
+	}
+	rawEvents, err := parseEventLog(e.MeasurementLog)
+	if err != nil {
+		return nil, fmt.Errorf("parsing measurement log: %v", err)
+	}
+	events, err = replayEvents(rawEvents, pcrs)
+	if err != nil {
+		return nil, fmt.Errorf("pcrs failed to replay")
+	}
+	return events, nil
+}
+
+type rawAttestationData struct {
```

Comment:
This seems TPM 1.2 specific?

Reply:
Yep, renaming to `rawQuoteInfo12` and adding links to the spec :)